Fix incorrect password behavior and message issue #14084

minagerges opened the pull request:
### What does it do?
Change the checks and the order of checks in checkIsBlocked method.

### Why is it needed?
When failed login is attempted N number of times, the account gets blocked as it was blocked by an admin, and displayed message is "login_blocked_admin" instead of "login_blocked_too_many_attempts"

### Related issue(s)/PR(s)
#14084

minagerges committed:
#14084

Fix failed login block behavior

minagerges committed:
#14084

Fix failed login block behavior (admin block message)

jonleverrier commented:
I've tested the process aspect of this PR;

1. After 5 failed login attempts I get the following error message "You have been blocked due to too many failed login attempts." vs "You have been blocked by an administrator."
2. The user is blocked by the value in `blocked by` and the `blocked checkbox` remains unticked vs the value in `blocked by` and the `blocked checkbox` ticked

So functionality works as described.

Mark-H commented on core/model/modx/processors/security/login.class.php:
```diff
@@ -140,23 +140,29 @@ public function checkIsBlocked() {
             $profile->save();
         }
-        if ($profile->get('blockeduntil') != 0 && $profile->get('blockeduntil') < time()) {
+        if ($this->user->passwordMatches($this->givenPassword)) {
```
This seems a bit odd to be added to a checkIsBlocked method - is it necessary to check the password here again?

opengeek replied:
More than odd; would this not destroy the ability to use external authentication via plugins?

minagerges replied:
No, it will not, it only resets "failedlogincount" and has nothing to do with the evaluation of blocked state. 
A successful login should reset "failedlogincount". Using an external authentication plugin will still pass, yet "failedlogincount" will be reset time based on next login.

opengeek replied:
With an external authentication plugin, the password will not be stored locally. So this does not seem like a valid solution. I will need to dig into and test this with a system using external authentication, but I can almost guarantee this will not work.

JoshuaLuckers replied:
Wouldn't it make more sense to reset the `failedlogincount` in `afterLogin()` instead of doing so in `checkIsBlocked()`?

minagerges replied:
@JoshuaLuckers, A successful login should reset `failedlogincount` **before** `checkIsBlocked()`, so it would bypass the `failedlogincount` block status _(as the user should not be blocked anymore based on the successful login)_.
@opengeek 's assumption is **only** valid if a user is already blocked because of `failedlogincount`, then successfully logs in via an external authentication plugin _(excluding Login extra)_. Only in such case, as I replied earlier, it will not reset the count and will be time based. Thus, it's a specific use case, based on the assumption of that (authentication plugin) incrementing  `failedlogincount` manually, as it's only incremented via the password validation (refer to `checkPassword()` logic).

minagerges commented:
The whole class is odd, and should br rewritten!
I explained everything in the issue, and the lines after the if statement
is clear, a successful password resets the counter, as per the common
behavior.
In the wrong place? Well yes, also explained that before!!!
Keeping changes to minimal, for a sensitive class.

I didn't have time to rewrite the whole class. May be in 2 weeks, after I'm
back from my vacation.

JoshuaLuckers commented:
A rewrite might be good for 3.x

sonicpunk commented:
So will this need to be rewritten or is it ok the way it is to merge?

Jako commented:
It should be ok for the moment, but it has to be checked.

sonicpunk commented:
@Jako I just installed it and tested the use-case and got the desired text output:
<img width="576" alt="screen shot 2018-10-13 at 12 25 36" src="https://user-images.githubusercontent.com/447453/46904355-31035480-cee3-11e8-9a57-375a6719dd88.png">

JoshuaLuckers commented on core/model/modx/processors/security/login.class.php:
```diff
@@ -140,23 +140,29 @@ public function checkIsBlocked() {
             $profile->save();
         }
-        if ($profile->get('blockeduntil') != 0 && $profile->get('blockeduntil') < time()) {
+        if ($this->user->passwordMatches($this->givenPassword)) {
```
I just ran some tests. If I try to login a few times with a wrong password I get the `login_blocked_too_many_attempts` message as expected.

If I login a few more times with a wrong password the message stays the same.
But when I login after `login_blocked_too_many_attempts` with the correct password I can't login because I'm blocked (as expected). 

When I try to login again with a wrong/correct password the message I receive is different (`login_blocked_error`). This happens because after the password matches `failedlogincount` is reset to 0.

An attacker can use this behaviour to retrieve a users password.

JoshuaLuckers replied:
@minagerges do you have time to check my comment?

minagerges replied:
@JoshuaLuckers please test based on PR commit https://github.com/modxcms/revolution/pull/14085/commits/3183308eb861149f43747c09367f7863c61a1f8e

minagerges replied:
@JoshuaLuckers please test against https://github.com/modxcms/revolution/pull/14085/commits/6f07f88772767c126c0e07231edba9a8a2c4e8b3 as latest change.

Fixed blocked occurring on `failedlogincount` + 1

Ibochkarev commented:
@minagerges is it possible to help us getting this merged?

minagerges committed:
Update login.class.php

minagerges reviewed (commented):
**Regression**: Correct password during block period changes returned message. 
https://github.com/modxcms/revolution/pull/14085#discussion_r243739086

**Change**: Correct password resets `failedlogincount` only after `blockeduntil` is past

minagerges committed:
Update login.class.php

Fix blocking is triggered on `failedlogincount` + 1
Caused by `checkPassword()`, which sets `failedlogincount` being triggered after `beforeLofin()`, which calls `checkIsBlocked`

minagerges committed:
Update login.class.php

Correct `failedlogincount` returned message logic to match https://github.com/modxcms/revolution/pull/14085/commits/9507d77bd1199ffb185e4b16df177344c414c676

minagerges committed:
Update login.class.php
